### ISSUE TITLE
PaginatedList decode error when query list

### ISSUE THREAD

**Issue by springlo** [aws-amplify/amplify-ios#860]
create `GraphQLRequest` :
```
extension GraphQLRequest where R == PaginatedList<Post> {
    static func listPublicPosts( _ anonymous: Bool = false, limit: Int? = nil, nextToken: String? = nil) -> GraphQLRequest {
        var variables: [String: Any] = [:]
        variables = [
            "filter": [
                "isPublic": [
                    "eq": true,
                ],
            ],
        ]
        if let limit = limit {
            variables["limit"] = limit
        }
        if let token = nextToken {
            variables["nextToken"] = token
        }
        return GraphQLRequest(apiName: anonymous ? API.apiName.api : API.apiName.authApi,
                              document: getPosts, variables: variables, responseType: PaginatedList<Post>.self, decodePath: "listPosts")
    }
}
```

and then request:
```Amplify.API.query(request: .listPublicPosts(anonymous, limit: 10))```

I got error:
```
DataStoreError: Error decoding the the sync metadata from the delta sync query result.
Recovery suggestion: The sync metadata should contain fields named `_deleted`, `_lastChangedAt` and `_version`.
Check your `syncPosts` query and make sure it returns the correct set of sync fields.
```

I know this is a `MutationSync` decoding throw, but my object `Post` has not  `_lastChangedAt` and `_version`.
So what should I do？ thx

**Comment by lawmicha** [comment]
Hi @springlo, 

`Amplify.API.query` shouldn't have produced that error message you see there, and should be coming from the sync engine in DataStore when performing a sync query. The error is that it failed to decode the response. Did you provision your API enabled for DataStore, ie. conflict resolution enabled? Does the sync query exist and are you able to perform a sync query operation from the AppSync console successfully?

**Comment by lawmicha** [comment]
Are you trying to use DataStore and APi together? if you use DataStore, the data is automatically synchronized to the local store provided by the DataStore plugin and you shouldn't have to use API plugin at all. However I do see that you have two APIs configured `.api` and `.authApi`. Some troubleshooting points
- If you have DataStore configured, it expects to use the first and API configured and we do not currently support multiple APIs configured at the same time. You can imagine, DataStore could take in configuration that indicates "which API by apiName to call"
- If you have a single API provisioned, it should be enabled for DataStore by enabling conflict resolution, and all calls should go through DataStore operations like `DataStore.query` instead of `API.query`

**Comment by springlo** [comment]
I need users to use some functions without logging in, so I request data like this.
I don’t use `DataStore `, can I use `Amplify.API` to request paging data?

**Comment by johnrbent** [comment]
I may have a solution to the problem...
I didn't know that the PaginatedList type even existed (where is that documented?) and it took me a while to figure out how to get the nextToken out of the response. So I just ended up creating my own type (replacing PaginatedList) to decode the response to:
```
public struct ItemsAndToken<M: Model>: Codable
{
    public let items: [M]
    public let nextToken: String?
}
```
and then for your request...

```
GraphQLRequest<ItemsAndToken<M>>(apiName: "[API_NAME]",
                                   document: documentString,
                                   variables: variables,
                                   responseType: ItemsAndToken<M>.self,
                                   decodePath: document.name)
```
 
The documentation could use some improvement in this area. Specifically around creating requests for more varied use cases. For example we have to dig into the SDK code to find how to create a GraphQLRequest when using multiple API or multi-auth. Figuring out how to decode paginated responses took considerably more time. The method suggested by the documentation to list Todo's seems like it would fail for >1000 Todo's. There's no suggestion of the implied 1,000 element pagination limit or how to retrieve the nextToken, or even that pagination is involved in that query.

I realize that this is a new technology and documentation will take some time. Just trying to provide some feedback that could help people like me who are learning backend processes for the first time.

**Comment by springlo** [comment]
@johnrbent  Thank you very much, I'm going to try the method you provided, it looks very effective.

**Comment by lawmicha** [comment]
hi @springlo, please see https://github.com/aws-amplify/amplify-ios/pull/1009 as it has been released in 1.6.0, this provides the API to make requests for list of data:

**Example following your original code snippet**
```swift
Amplify.API.query(.paginatedList(Post.self, where: Post.keys.isPublic == true, limit: limit)) { result in
   switch result {
   case .success(let graphQLResponse):
       switch graphqlResponse {
          case .success(let posts):
          case .failure(let graphQLErrorResponse):
       }
   case .failure(let apiError):
   }
}
```
You may still need to perform a bit of custom logic to copy of the `GraphQLRequest` and add the `apiName` since it is immutable. For example, 
```swift
let request = GraphQLRequest.paginatedList( ...)
let newRequest = copyToNewGraphQLRequest(request, anonymous, apiName)
```

Once you have performed the request, you can check if there is a next page and get the page:
```
posts.hasNextPage()
posts.getNextPage()
```
Let us know if you have any questions
